File: app/src/main/java/igrek/todotree/service/access/DatabaseLock.java
package igrek.todotree.service.access;


import igrek.todotree.domain.treeitem.AbstractTreeItem;
import igrek.todotree.exceptions.DatabaseLockedException;
import igrek.todotree.info.logger.Logger;
import igrek.todotree.info.logger.LoggerFactory;
import igrek.todotree.service.preferences.Preferences;
import igrek.todotree.service.preferences.PropertyDefinition;

public class DatabaseLock {
	
	private boolean locked = true;
	
	private Logger logger = LoggerFactory.INSTANCE.getLogger();
	private AccessLogService accessLogService;
	
	public DatabaseLock(Preferences preferences, AccessLogService accessLogService) {
		locked = preferences.getValue(PropertyDefinition.lockDB, Boolean.class);
		this.accessLogService = accessLogService;
	}
	
	public boolean isLocked() {
		return locked;
	}
	
	public void setLocked(boolean locked) {
		this.locked = locked;
	}
	
	public boolean unlockIfLocked(AbstractTreeItem item) {
		if (isLocked()) {
			setLocked(false);
			accessLogService.logDBUnlocked(item);
			logger.debug("Database unlocked.");
			return true;
		}
		return false;
	}
	
	public void assertUnlocked() throws DatabaseLockedException {
		if (locked)
			throw new DatabaseLockedException();
	}
}

File: app/src/main/java/igrek/todotree/service/access/DatabaseLock.kt
package igrek.todotree.service.access

import igrek.todotree.domain.treeitem.AbstractTreeItem
import igrek.todotree.exceptions.DatabaseLockedException
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import igrek.todotree.settings.SettingsState

class DatabaseLock(
    settingsState: LazyInject<SettingsState> = appFactory.settingsState,
    accessLogService: LazyInject<AccessLogService> = appFactory.accessLogService,
) {
    private val accessLogService by LazyExtractor(accessLogService)

    var isLocked = true
    private val logger = LoggerFactory.logger

    fun unlockIfLocked(item: AbstractTreeItem?): Boolean {
        if (isLocked) {
            isLocked = false
            accessLogService.logDBUnlocked(item)
            logger.debug("Database unlocked.")
            return true
        }
        return false
    }

    @Throws(DatabaseLockedException::class)
    fun assertUnlocked() {
        if (isLocked) throw DatabaseLockedException()
    }

    init {
        isLocked = settingsState.get().lockDB
    }
}